### PR TITLE
Qcxms

### PR DESCRIPTION
**Pull request by zargham-ahmad**
This PR includes the following: #419 

- added qcxms_neutral_run tool, which outputs three collections (apart from the additional output) which are further used by qcxms_prod_run. 
- added qcxms_prod_run which uses three collections from qcxms_neutral_run and outputs two files one is log file and the other is result.jdx file

TODO:
- [x] update the qcxms version to 5.2.1
- [ ]  option to use pqcxms command in qcxms_prod_run (cannot use until the next release because of the prod flag issue https://github.com/qcxms/QCxMS/pull/80)
- [x]  add plotms as a requirement after the package gets on conda

### CODE REVIEW

**Comment by hechth** [comment]
@zargham-ahmad please add a description for this PR shortly outlining the tool added and what was done for documentation purposes.

**Comment by zargham-ahmad** [comment]
> @zargham-ahmad please add a description for this PR shortly outlining the tool added and what was done for documentation purposes.

done

File: tools/qcxms/qcxms_prod_run.xml [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+    </macros>
+
+    <expand macro="creator"/>
+
+    <edam_topics>
+        <edam_topic>topic_3332</edam_topic>
+    </edam_topics>
+    <edam_operations>
+        <edam_operation>operation_0297</edam_operation>
+    </edam_operations>
```
Comment (hechth) [inline tools/qcxms/qcxms_prod_run.xml:15]: Do we have a bio.tools entry for the package?

File: tools/qcxms/qcxms_prod_run.xml [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+#set folder2_files = str("', '").join([str($f) for $f in $start_files])
+#set folder3_files = str("', '").join([str($f) for $f in $xyz_files])
+
+#set names = str("', '").join([str($f.name) for $f in $xyz_files])
+names = '$names'
+folder_names = [x.split("_")[0] for x in names]
+
+folder1_files = '$folder1_files'
+folder2_files = '$folder2_files'
+folder3_files = '$folder3_files'
```
Comment (hechth) [inline tools/qcxms/qcxms_prod_run.xml:47]: Variable names could be improved, like "input_config_files" etc.

File: tools/qcxms/qcxms_prod_run.xml [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+
+# Create a new output folder to store the result
+output_path = 'TMPQCXMS'
+os.makedirs(output_path, exist_ok=True)
+
+for fname, filename1, filename2, filename3 in zip(folder_names, folder1_files, folder2_files, folder3_files):
```
Comment (hechth) [inline tools/qcxms/qcxms_prod_run.xml:53]: Same here - variable names should be more concise to allows easy maintenance later on and to make the code easier to read.

File: tools/qcxms/.shed.yml [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+  The program was originally developed to calculate Electron Ionization (EI) mass spectra, 
+  in which a (typically 70 eV) electron beam is focused on a molecule in order to create an open-shell radical ion (uneven number of valence electrons).
+  This process not only ionizes the molecule, but simultaneously increases the internal energy of the species, which in turn leads to bond breaking, 
+  fragmentation, rearrangement, etc of the ion."
+auto_tool_repositories:
+  name_template: "{{ tool_id }}"
+  description_template: "{{ tool_name }} tool from the QCxMS package"
+suite:
+  name: suite_qcxms
+  description: tools from QCxMS are used for molecular geometry optimization and in silico calculation of mass spectra using quantum chemistry
```
Comment (hechth) [inline tools/qcxms/.shed.yml:21]: Tools are only for spectrum prediction

File: tools/qcxms/qcxms_neutral_run.xml [hunk 5]
```diff
@@ -0,0 +1,145 @@
+<tool id="qcxms_neutral_run" name="QCxMS neutral run" version="@TOOL_VERSION@+galaxy0" python_template_version="3.5" profile="21.05">
```
Comment (bgruening) [inline tools/qcxms/qcxms_neutral_run.xml:1]:
```suggestion
<tool id="qcxms_neutral_run" name="QCxMS neutral run" version="@TOOL_VERSION@+galaxy0" profile="21.05">
```

File: tools/qcxms/qcxms_neutral_run.xml [hunk 6]
```diff
@@ -0,0 +1,145 @@
...
+    <edam_operations>
+        <edam_operation>operation_0297</edam_operation>
+    </edam_operations>
+
+    <requirements>
+        <requirement type="package" version="@TOOL_VERSION@">qcxms</requirement>
+    </requirements>
+
+    <command detect_errors="exit_code"><![CDATA[
+        ln $mol molecule.xyz &&
```
Comment (bgruening) [inline tools/qcxms/qcxms_neutral_run.xml:22]: ln -s
```suggestion
        ln -s '$mol' molecule.xyz &&
```

File: tools/qcxms/qcxms_neutral_run.xml [hunk 7]
```diff
@@ -0,0 +1,145 @@
...
+        cat qcxms.in &&
+        qcxms -i molecule.xyz >> $log &&
+        qcxms -i molecule.xyz >> $log &&
```
Comment (bgruening) [inline tools/qcxms/qcxms_neutral_run.xml:25]:
```suggestion
        qcxms -i molecule.xyz >> '$log' &&
        qcxms -i molecule.xyz >> '$log' &&
```

File: tools/qcxms/qcxms_prod_run.xml [hunk 8]
```diff
@@ -0,0 +1,90 @@
...
+    <edam_operations>
+        <edam_operation>operation_0297</edam_operation>
+    </edam_operations>
+
+    <requirements>
+        <requirement type="package" version="@TOOL_VERSION@">qcxms</requirement>
+    </requirements>
+
+    <command detect_errors="exit_code"><![CDATA[
+        python3 ${create_folder} &&
```
Comment (bgruening) [inline tools/qcxms/qcxms_prod_run.xml:22]:
```suggestion
        python3 '${create_folder}' &&
```

File: tools/qcxms/qcxms_prod_run.xml (hunk 3 above)
Comment (zargham-ahmad) [reply tools/qcxms/qcxms_prod_run.xml]: fixed in [77886de](https://github.com/RECETOX/galaxytools/pull/421/commits/77886de66a13a179bbbb88bbd5ade5689683dadb)

File: tools/qcxms/qcxms_prod_run.xml (hunk 2 above)
Comment (zargham-ahmad) [reply tools/qcxms/qcxms_prod_run.xml]: fixed in [77886de](https://github.com/RECETOX/galaxytools/pull/421/commits/77886de66a13a179bbbb88bbd5ade5689683dadb)

File: tools/qcxms/qcxms_prod_run.xml (hunk 1 above)
Comment (zargham-ahmad) [reply tools/qcxms/qcxms_prod_run.xml]: no

File: tools/qcxms/qcxms_neutral_run.xml [hunk 9]
```diff
@@ -0,0 +1,145 @@
...
+            <output name="log">
+                <assert_contents>
+                    <has_size value="10518" delta="300"/>
+                </assert_contents>
+            </output>
+        </test>
+    </tests>
+
+    <help><![CDATA[]]>
+    </help>
```
[the remark this answers is not in the document]
Comment (zargham-ahmad) [reply tools/qcxms/qcxms_neutral_run.xml]: updated in [bc5f153](https://github.com/RECETOX/galaxytools/pull/421/commits/bc5f15331788dacd0de6a7b45fe15db0c8b66000)

**Comment by zargham-ahmad** [comment]
qcxms is failing for more trajectories with the following error but the compiled binary from qcxms github seems to be working fine

```
free(): invalid pointer

Program received signal SIGABRT: Process abort signal.

Backtrace for this error:
#0  0x7f6eece94fcf in ???
#1  0x7f6eecee3d3c in ???
#2  0x7f6eece94f31 in ???
#3  0x7f6eece7f471 in ???
#4  0x7f6eeced833f in ???
#5  0x7f6eeceed6b9 in ???
#6  0x7f6eeceef443 in ???
#7  0x7f6eecef1d9e in ???
#8  0x560b5c56e05e in ???
#9  0x560b5c535ef5 in ???
#10  0x560b5c533060 in ???
#11  0x7f6eece801c9 in ???
#12  0x7f6eece80284 in ???
#13  0x560b5c533090 in ???
```

**Comment by zargham-ahmad** [comment]
Hi, the conda forge version of the qcxms is failing with the below error. @bgruening @JayTheDog, do you have any idea? 

free(): invalid pointer

Program received signal SIGABRT, Aborted.
__GI_raise (sig=sig@entry=6) at ../sysdeps/unix/sysv/linux/raise.c:50
50      ../sysdeps/unix/sysv/linux/raise.c: No such file or directory.
(gdb) bt
#0  __GI_raise (sig=sig@entry=6) at ../sysdeps/unix/sysv/linux/raise.c:50
#1  0x00007ffff56fe859 in __GI_abort () at abort.c:79
#2  0x00007ffff576926e in __libc_message (action=action@entry=do_abort, fmt=fmt@entry=0x7ffff5893298 "%s\n")
    at ../sysdeps/posix/libc_fatal.c:155
#3  0x00007ffff57712fc in malloc_printerr (str=str@entry=0x7ffff58914c1 "free(): invalid pointer") at malloc.c:5347
#4  0x00007ffff5772b2c in _int_free (av=<optimized out>, p=<optimized out>, have_lock=0) at malloc.c:4173
#5  0x000055555559305f in __qcxms_write_fragments_MOD_manage_fragments ()
#6  0x000055555555aef6 in MAIN__ ()
(gdb) continue
Continuing.

Program received signal SIGABRT: Process abort signal.

Backtrace for this error:
#0  0x7ffff571f08f in ???
        at /build/glibc-BHL3KM/glibc-2.31/signal/../sysdeps/unix/sysv/linux/x86_64/sigaction.c:0
#1  0x7ffff571f00b in __GI_raise
        at ../sysdeps/unix/sysv/linux/raise.c:51
#2  0x7ffff56fe858 in __GI_abort
        at /build/glibc-BHL3KM/glibc-2.31/stdlib/abort.c:79
#3  0x7ffff576926d in __libc_message
        at ../sysdeps/posix/libc_fatal.c:155
#4  0x7ffff57712fb in malloc_printerr
        at /build/glibc-BHL3KM/glibc-2.31/malloc/malloc.c:5347
#5  0x7ffff5772b2b in _int_free
        at /build/glibc-BHL3KM/glibc-2.31/malloc/malloc.c:4173
#6  0x55555559305e in ???
#7  0x55555555aef5 in ???
#8  0x555555558060 in ???
#9  0x7ffff5700082 in __libc_start_main
        at ../csu/libc-start.c:308
#10  0x555555558090 in ???

Program received signal SIGABRT, Aborted.
__GI_raise (sig=sig@entry=6) at ../sysdeps/unix/sysv/linux/raise.c:50

**Comment by JayTheDog** [comment]
> Hi, the conda forge version of the qcxms is failing with the below error. @bgruening @JayTheDog, do you have any idea?
> 
> free(): invalid pointer
> 
> Program received signal SIGABRT, Aborted. __GI_raise (sig=sig@entry=6) at ../sysdeps/unix/sysv/linux/raise.c:50 50 ../sysdeps/unix/sysv/linux/raise.c: No such file or directory. (gdb) bt #0 __GI_raise (sig=sig@entry=6) at ../sysdeps/unix/sysv/linux/raise.c:50 #1 0x00007ffff56fe859 in __GI_abort () at abort.c:79 #2 0x00007ffff576926e in __libc_message (action=action@entry=do_abort, fmt=fmt@entry=0x7ffff5893298 "%s\n") at ../sysdeps/posix/libc_fatal.c:155 #3 0x00007ffff57712fc in malloc_printerr (str=str@entry=0x7ffff58914c1 "free(): invalid pointer") at malloc.c:5347 #4 0x00007ffff5772b2c in _int_free (av=, p=, have_lock=0) at malloc.c:4173 #5 0x000055555559305f in **qcxms_write_fragments_MOD_manage_fragments () #6 0x000055555555aef6 in MAIN** () (gdb) continue Continuing.
> 
> Program received signal SIGABRT: Process abort signal.
> 
> Backtrace for this error: #0 0x7ffff571f08f in ??? at /build/glibc-BHL3KM/glibc-2.31/signal/../sysdeps/unix/sysv/linux/x86_64/sigaction.c:0 #1 0x7ffff571f00b in __GI_raise at ../sysdeps/unix/sysv/linux/raise.c:51 #2 0x7ffff56fe858 in __GI_abort at /build/glibc-BHL3KM/glibc-2.31/stdlib/abort.c:79 #3 0x7ffff576926d in __libc_message at ../sysdeps/posix/libc_fatal.c:155 #4 0x7ffff57712fb in malloc_printerr at /build/glibc-BHL3KM/glibc-2.31/malloc/malloc.c:5347 #5 0x7ffff5772b2b in _int_free at /build/glibc-BHL3KM/glibc-2.31/malloc/malloc.c:4173 #6 0x55555559305e in ??? #7 0x55555555aef5 in ??? #8 0x555555558060 in ??? #9 0x7ffff5700082 in __libc_start_main at ../csu/libc-start.c:308 #10 0x555555558090 in ???
> 
> Program received signal SIGABRT, Aborted. __GI_raise (sig=sig@entry=6) at ../sysdeps/unix/sysv/linux/raise.c:50

At what point does this happen? I currently don't have the capability to check whether this problem occurs only for some trajectories, or for the entirety of the runs. [@gorges](https://github.com/gorges97) has the capabilities for this, we should make him part of this conversation

**Comment by hechth** [comment]
@zargham-ahmad you mentioned that it works with a single trajectory but not multiple, is that right?

**Comment by hechth** [comment]
@JayTheDog @bgruening @gorges97 I think it is related to the compilation for conda-forge. I ran now 3 trajectories locally and the tool peaked at 61GB memory consumption using the binary from conda forge - probably some optimizations are not available or the compiler flags are not passed correctly?

**Comment by bgruening** [comment]
https://github.com/conda-forge/qcxms-feedstock/blob/main/recipe/build.sh

What is missing here?

File: tools/qcxms/qcxms_prod_run.xml [hunk 10]
```diff
@@ -0,0 +1,86 @@
...
+
+    <outputs>
+        <data name="msp_output" format="msp" from_work_dir="simulated_spectra.msp" label="simulated_spectra.msp generated by ${tool.name} on ${on_string}"/>
+        <data name="log" format="txt" label="logfile of ${tool.name} on ${on_string}">
+            <filter>store_extended_output</filter>
+        </data>
+    </outputs>
+
+    <tests>
+    </tests>
```
Comment (hechth) [inline tools/qcxms/qcxms_prod_run.xml:70]:
```suggestion
    <tests>
        <test expect_failure="true"/>
    </tests>
```

**Review by hechth — APPROVED** [review]

File: tools/qcxms/qcxms_prod_run.xml [hunk 11]
```diff
@@ -0,0 +1,87 @@
...
+    <macros>
+        <import>macros.xml</import>
+    </macros>
+    <expand macro="edam"/>
+    <expand macro="creator"/>
+    <expand macro="requirements"/>
+
+    <command detect_errors="exit_code"><![CDATA[
+        python3 '${create_folder_structure}' &&
+        find TMPQCXMS/*/ -type d | xargs -I {} -P 4 sh -c 'cd {} && /qcxms_bin/qcxms --prod >> $log' &&
```
Comment (hechth) [inline tools/qcxms/qcxms_prod_run.xml:13]:
```suggestion
        find TMPQCXMS/*/ -type d | xargs -I {} -P ${GALAXY_SLOTS} sh -c 'cd {} && /qcxms_bin/qcxms --prod >> $log' &&
```

**Pull request merged**